### PR TITLE
Use beginning of first paragraph for "tooltip"  by Sphinx Gallery extension

### LINKED ISSUE #2657: Link Checker Report on 2023-09-03
## Summary

| Status        | Count |
|---------------|-------|
| 🔍 Total      | 20552 |
| ✅ Successful | 5334  |
| ⏳ Timeouts   | 4     |
| 🔀 Redirected | 0     |
| 👻 Excluded   | 15201 |
| ❓ Unknown    | 0     |
| 🚫 Errors     | 13    |

### Errors in documentation/dev/gallery/index.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in repository/pygmt/src/makecpt.py

* [TIMEOUT] [http://soliton.vm.bytemark.co.uk/pub/cpt-city/](http://soliton.vm.bytemark.co.uk/pub/cpt-city/) | Timeout

### Errors in documentation/dev/api/generated/pygmt.makecpt.html

* [TIMEOUT] [http://soliton.vm.bytemark.co.uk/pub/cpt-city/](http://soliton.vm.bytemark.co.uk/pub/cpt-city/) | Timeout

### Errors in documentation/dev/api/generated/pygmt.Figure.show.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in documentation/dev/api/generated/pygmt.Figure.colorbar.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Failed: Network error: dns error: no record found for Query { name: Name("docs.gmt-china.ou4bps3zwvpehennzi3rrspmbh.dx.internal.cloudapp.net."), query_type: AAAA, query_class: IN }

### Errors in documentation/dev/api/generated/pygmt.project.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in documentation/dev/api/generated/pygmt.datasets.load_earth_relief.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in repository/examples/gallery/lines/roads.py

* [503] [http://www2.census.gov/geo/tiger/TIGER2015/PRISECROADS/tl_2015_15_prisecroads.zip](http://www2.census.gov/geo/tiger/TIGER2015/PRISECROADS/tl_2015_15_prisecroads.zip) | Failed: Network error: Service Unavailable

### Errors in documentation/dev/api/generated/pygmt.Figure.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in documentation/dev/api/generated/pygmt.Figure.plot.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in documentation/dev/api/generated/pygmt.Figure.text.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in documentation/dev/api/generated/pygmt.Figure.basemap.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Failed: Network error: dns error: no record found for Query { name: Name("docs.gmt-china.ou4bps3zwvpehennzi3rrspmbh.dx.internal.cloudapp.net."), query_type: AAAA, query_class: IN }

### Errors in documentation/dev/api/generated/pygmt.grd2cpt.html

* [TIMEOUT] [http://soliton.vm.bytemark.co.uk/pub/cpt-city/](http://soliton.vm.bytemark.co.uk/pub/cpt-city/) | Timeout

### Errors in documentation/dev/api/generated/pygmt.Figure.grdimage.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in documentation/dev/api/generated/pygmt.grdtrack.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)

### Errors in repository/pygmt/src/grd2cpt.py

* [TIMEOUT] [http://soliton.vm.bytemark.co.uk/pub/cpt-city/](http://soliton.vm.bytemark.co.uk/pub/cpt-city/) | Timeout

### Errors in documentation/dev/api/generated/pygmt.Figure.shift_origin.html

* [ERR] [https://docs.gmt-china/](https://docs.gmt-china/) | Cached: Error (cached)
[Full Github Actions output](https://github.com/GenericMappingTools/pygmt/actions/runs/6064445782?check_suite_focus=true)

### PR DESCRIPTION
**Pull request by yvonnefroehlich**
**Description of proposed changes**

This PR aims to fix an error regarding the link to the GMT China documentation in the gallery example "Cross-section along a transect".

It seems like empty lines within the introduction part of a gallery example impact on what is used by the Sphinx-Gallery extension as text for `tooltip`.
Adding an empty line after the heading changes the text used for `tooltip` to the beginning of the first paragraph. Without any empty line within the introduction part the heading of the gallery example is used for `tooltip`.

_(0) Main branch_  -  `view-source:https://www.pygmt.org/dev/gallery/index.html`
```
</div><div class="sphx-glr-thumbcontainer" tooltip="This example is orientated on an example in the GMT/China documentation: https://docs.gmt-china..."><img alt="" src="../_images/sphx_glr_cross_section_thumb.png" />
<p><a class="reference internal" href="images/cross_section.html#sphx-glr-gallery-images-cross-section-py"><span class="std std-ref">Cross-section along a transect</span></a></p>
  <div class="sphx-glr-thumbnail-title">Cross-section along a transect</div>
```

_(1) Empty line after heading_ 
```
</div><div class="sphx-glr-thumbcontainer" tooltip="pygmt.project and pygmt.grdtrack can be used to focus on a quantity and its variation along a d..."><img alt="Cross-section along a transect" src="../_images/sphx_glr_cross_section_thumb.png" />
<p><a class="reference internal" href="images/cross_section.html#sphx-glr-gallery-images-cross-section-py"><span class="std std-ref">Cross-section along a transect</span></a></p>
  <div class="sphx-glr-thumbnail-title">Cross-section along a transect</div>
```

_(2) No empty lines_
```
</div><div class="sphx-glr-thumbcontainer" tooltip="Cross-section along a transect"><img alt="Cross-section along a transect" src="../_images/sphx_glr_cross_section_thumb.png" />
<p><a class="reference internal" href="images/cross_section.html#sphx-glr-gallery-images-cross-section-py"><span class="std std-ref">Cross-section along a transect</span></a></p>
  <div class="sphx-glr-thumbnail-title">Cross-section along a transect</div>
```

_Output_
| (0) main branch | (1) empty line after heading | (2) no empty lines |
|---|---|---|
| crosssection_tooltip_bug | crosssection_tooltip_fix | crosssection_tooltip_heading |
| crosssection_intro_bug | crosssection_intro_fix | crosssection_intro_heading |

So far, this PR contains version (1). Maybe/ if possible we should choose one version and update all gallery examples corresponding to it?

**Fixes**: #2657

**Preview**: 
- [x] Gallery examples: view-source:https://pygmt-dev--2659.org.readthedocs.build/en/2659/gallery/index.html
- [x] Tutorials: view-source:https://pygmt-dev--2659.org.readthedocs.build/en/2659/tutorials/index.html
- [x] Intros: view-source:https://pygmt-dev--2659.org.readthedocs.build/en/2659/get_started/index.html
- [x] Projections: no changes needed

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by seisman** [comment]
> It seems like empty lines within the introduction part of a gallery example impact on what is used by the Sphinx-Gallery extension as text for tooltip.
> Adding an empty line after the heading changes the text used for tooltip to the beginning of the first paragraph. Without any empty line within the introduction part the heading of the gallery example is used for tooltip.

Nice debugging! 

I just checked the Sphinx-Gallery source code. The text shown in tooltips is called `intro` in Sphinx-Gallery. Looking at the source code (https://github.com/sphinx-gallery/sphinx-gallery/blob/c1edfbbfaae695e7a3e239de370f8bd24457ab9d/sphinx_gallery/gen_rst.py#L263), for each example script, it first splits the file content into paragraphs, using `\n\n` as the separator:
```python
paragraphs = docstring.lstrip().split("\n\n")
```
Then, it uses the 2nd paragraph as `intro` (if available) and truncates it to <95 chars. If there is only one "paragraph", the title will be used in the "tooltip".
```python
    # Use the title if no other paragraphs are provided
    intro_paragraph = title if len(paragraphs) < 2 else paragraphs[1]
    # Concatenate all lines of the first paragraph and truncate at 95 chars
    intro = re.sub("\n", " ", intro_paragraph)
    intro = _sanitize_rst(intro)
    if len(intro) > 95:
        intro = intro[:95] + "..."
```
This clearly explains why an empty line below the "title" matters.

I think we should always add an empty line below the title. Maybe we should also report the "issue" to the upstream Sphinx-Gallery repository?

**Comment by yvonnefroehlich** [comment]
> I think we should always add an empty line below the title.

That's fine with me. Do we want to include this in this PR, or keep the change for the "cross section" example separately and start a new PR for the remaining examples?

**Comment by seisman** [comment]
Updating all examples in this PR sounds good to me.

**Comment by yvonnefroehlich** [comment]
> Maybe we should also report the "issue" to the upstream Sphinx-Gallery repository?

Yes, I feel this would be fair. However, in principle, it makes sense to limit the text used for `tooltip` to a maximum of characters. So, probably any kind of exception is needed for the case "a link is present and potentially cut off"?

**Comment by seisman** [comment]
> > Maybe we should also report the "issue" to the upstream Sphinx-Gallery repository?
> 
> Yes, I feel this would be fair. However, in principle, it makes sense to limit the text used for `tooltip` to a maximum of characters. So, probably any kind of exception is needed for the case "a link is present and potentially cut off"?

It makes sense to truncate the text for `tooltip`, so that's not the main issue. I think the issue is, that when there is no blank line below the title, Sphinx-Gallery "incorrectly" takes the 2nd paragraph as the text for `tooltip`.

File: examples/tutorials/basics/plot.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 """
 Plotting data points
---------------------
+====================
```
Comment (seisman) [inline examples/tutorials/basics/plot.py:3]: It seems we're inconsistently using `----` or `=====` below the title.

Comment (yvonnefroehlich) [reply examples/tutorials/basics/plot.py:3]: In gallery examples we are nearly all most using `----` below the title. But for the tutorials and intros, we are using `====` below the title and `----` below the headings of the subsections (if present). For consistency we maybe / probably should update the examples to have `====`  below the title?

Comment (seisman) [reply examples/tutorials/basics/plot.py:3]: Let's focus on the empty line in the PR, and decide to use `---` or `===` later.

Comment (yvonnefroehlich) [reply examples/tutorials/basics/plot.py:3]: Yes, it makes sense to handle these two aspects separately, as this title underline thing is unrelated to the `tooltip` issue.

**Comment by yvonnefroehlich** [comment]
> > > Maybe we should also report the "issue" to the upstream Sphinx-Gallery repository?
> > 
> > 
> > Yes, I feel this would be fair. However, in principle, it makes sense to limit the text used for `tooltip` to a maximum of characters. So, probably any kind of exception is needed for the case "a link is present and potentially cut off"?
> 
> It makes sense to truncate the text for `tooltip`, so that's not the main issue. I think the issue is, that when there is no blank line below the title, Sphinx-Gallery "incorrectly" takes the 2nd paragraph as the text for `tooltip`.

Ah. Yes, this is unexpected, since it is technically not required to have a blank line below the title.

File: examples/gallery/images/cross_section.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 """
 Cross-section along a transect
-==============================
+------------------------------
```
Comment (seisman) [inline examples/gallery/images/cross_section.py:3]: Revert this change and address it in a separate PR?
```suggestion
==============================
```

File: examples/tutorials/basics/plot.py (hunk 1 above)
Comment (seisman) [inline examples/tutorials/basics/plot.py:3]:
```suggestion
--------------------
```

**Comment by seisman** [comment]
@yvonnefroehlich It seems this PR is almost ready for review, right?

**Review by seisman — APPROVED** [review]

**Review by michaelgrund — APPROVED** [review]
